stdlib/asm/math/poly512.masm

```diff
@@ -0,0 +1,638 @@
+
+        movup.6
+        add.1
+        movdn.6
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+
+        push.0.0.0.0
+    end
+
+    push.0
+    dropw
+    dropw
+    dropw
+
+    push.env.locaddr.127
+end
+
+# Given one polynomial of degree 512 on stack as absolute memory addresses,
+# this routine negates each coefficient of that polynomial.
+#
+# Imagine, polynomial f
+#
+# g = -f, can be computed as
+#
+# [(-f[i]) % Q for i in range(512)] | Q = 12289 ( = Falcon Digital Signature Algorithm's Prime Number )
+#
+# Input stack state :
+#
+# [f_start_addr, ...]
+#
+# - f_addr`i` -> f[ (i << 2) .. ((i+1) << 2) ), address holding four consecutive coefficients
+# - f_addr0 -> f_start_addr
+#
+# Output stack state :
+#
+# [g_start_addr, ...]
+#
+# - g_addr`i` is the address of four elements | g[ (i << 2) .. ((i+1) << 2) ) & i ∈ [0..128)
+#
+# Consecutive 127 memory addresses can be computed from starting memory address ( living on stack top ) by 
+# continuing to apply `INCR` ( = add.1 ) instruction on previous absolute memory address.
+#
+# Note, input memory addresses are considered to be read-only, they are not mutated.
+export.neg_zq.128
+    push.env.locaddr.127
+    push.0.0.0.0
+
+    repeat.128
+        dup.5
+        loadw.mem
+
+        exec.neg_word
+
+        dup.4
+        storew.mem
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+    end
+
+    dropw
+    drop
+    drop
+
+    push.env.locaddr.127
+end
+
+# Given two polynomials of degree 512 on stack as absolute memory addresses,
+# this routine subtracts second polynomial from first one.
+#
+# Imagine, two polynomials f, g
+#
+# h = f - g, can be computed as
+#
+# [(f[i] - g[i]) % Q for i in range(512)] | Q = 12289 ( = Falcon Digital Signature Algorithm's Prime Number )
+#
+# Input stack state :
+#
+# [f_start_addr, g_start_addr, ...]
+#
+# - {f, g}_addr`i` -> {f, g}[ (i << 2) .. ((i+1) << 2) ), address holding four consecutive coefficients
+# - {f, g}_addr0 -> {f, g}_start_addr
+#
+# Output stack state :
+#
+# [h_start_addr, ...]
+#
+# - h_addr`i` is the address of four elements | h[ (i << 2) .. ((i+1) << 2) ) & i ∈ [0..128)
+#
+# Consecutive 127 memory addresses can be computed from starting memory address ( living on stack top ) by 
+# continuing to apply `INCR` ( = add.1 ) instruction on previous absolute memory address.
+#
+# Note, input memory addresses are considered to be read-only, they are not mutated.
+export.sub_zq.256
+    push.env.locaddr.127
+
+    repeat.128
+        dup.1
+        dup.1
+        pop.mem
+
+        add.1
+        swap
+        add.1
+        swap
+    end
+
+    drop
+    drop
+
+    exec.neg_zq
+
+    push.env.locaddr.255
+    push.0.0.0.0
+
+    repeat.128
+        dup.5
+        loadw.mem
+
+        dup.4
+        storew.mem
+
+        movup.5
+        add.1
+        movdn.5
+
+        movup.4
+        add.1
+        movdn.4
+    end
```

Comment:
If I understood how this works correctly, by the time we get here, we already have negated polynomial $f$. Can we not just apply `add_zq` procedure directly to the results of the negated polynomial without this extra copy?

Reply:
We need these memcpy(s) because of overlapping memory regions, where input polynomial lives and called function will also attempt to store output polynomial. But this is avoided starting from commit f351eaf8f7b0cf85f942a1eea39db02b852a5f32